packages/libs/error-stack/src/fmt/mod.rs

```diff
@@ -0,0 +1,1104 @@
+        } else {
+            queue.push(*frame);
+        }
+    }
+
+    (result, queue)
+}
+
+fn debug_context(frame: &Frame, context: &dyn Context, alternate: bool) -> (Lines, Line) {
+    let loc = frame.location();
+    let context = context
+        .to_string()
+        .lines()
+        .map(ToOwned::to_owned)
+        .enumerate()
+        .map(|(idx, value)| {
+            if idx == 0 {
+                Line::new().push(Instruction::Value {
+                    value,
+                    style: Style::new().bold(),
+                })
+            } else {
+                Line::new().push(Instruction::Value {
+                    value,
+                    style: Style::new(),
+                })
+            }
+        })
+        .enumerate()
+        .map(|(idx, line)| {
+            if idx == 0 {
+                line
+            } else {
+                line.push(Instruction::Indent {
+                    group: false,
+                    visible: true,
+                    spacing: true,
+                    minimal: true,
+                })
+            }
+        })
+        .collect();
+
+    let loc = Line::new()
+        .push(Instruction::Value {
+            value: if alternate {
+                format!("{loc:?}")
+            } else {
+                loc.to_string()
+            },
+            style: Style::new().gray(),
+        })
+        .push(Instruction::Symbol(Symbol::Location));
+
+    (context, loc)
+}
+
+struct Opaque(usize);
+
+impl Opaque {
+    const fn new() -> Self {
+        Self(0)
+    }
+
+    fn increase(&mut self) {
+        self.0 += 1;
+    }
+
+    fn render(self) -> Option<Line> {
+        match self.0 {
+            0 => None,
+            1 => Some(Line::new().push(Instruction::Value {
+                value: "1 additional opaque attachment".to_owned(),
+                style: Style::new(),
+            })),
+            n => Some(Line::new().push(Instruction::Value {
+                value: format!("{n} additional opaque attachments"),
+                style: Style::new(),
+            })),
+        }
+    }
+}
+
+fn debug_attachments(
+    loc: Option<Line>,
+    ctx: &mut HookContextImpl,
+    last: bool,
+    frames: Vec<&Frame>,
+) -> Lines {
+    let mut opaque = Opaque::new();
+
+    // evaluate all frames to their respective values, will call all hooks with the current context
+    let (next, defer): (Vec<_>, _) = frames
+        .into_iter()
+        .map(|frame| match frame.kind() {
+            FrameKind::Context(_) => unreachable!(),
+            FrameKind::Attachment(AttachmentKind::Opaque(_)) => {
+                #[cfg(all(nightly, feature = "experimental"))]
+                if let Some(debug) = frame.request_ref::<DebugDiagnostic>() {
+                    for snippet in debug.snippets() {
+                        ctx.cast::<()>().add_snippet(snippet.clone());
+                    }
+
+                    return Some(debug.output().clone());
+                }
```

Comment:
Does it make a difference if `ctx.cast::<()>` would be changed to `ctx.cast::<DebugDiagnostic>`?

Reply:
Nope, not at all, but it would make it clearer! Good catch.